multiaddr/multiaddr.py

```diff
@@ -100,35 +189,35 @@ def decapsulate(self, other):
         For example:
             /ip4/1.2.3.4/tcp/80 decapsulate /ip4/1.2.3.4 = /tcp/80
         """
-        s1 = str(self)
-        s2 = str(other)
+        s1 = self.to_bytes()
+        s2 = Multiaddr(other).to_bytes()
         try:
             idx = s1.rindex(s2)
         except ValueError:
             # if multiaddr not contained, returns a copy
-            return copy(self)
+            return Multiaddr(self)
         return Multiaddr(s1[:idx])
 
     def value_for_protocol(self, proto):
-        """Return the value (if any) following the specified protocol."""
-        if not isinstance(proto, protocols.Protocol):
-            if isinstance(proto, int):
-                proto = protocols.protocol_with_code(proto)
-            elif isinstance(proto, six.string_types):
-                proto = protocols.protocol_with_name(proto)
-            else:
-                raise TypeError("Protocol object, name or code expected, got {0!r}".format(proto))
-
-        for proto2, codec, part in bytes_iter(self.to_bytes()):
-            if proto2 == proto:
-                if codec.SIZE != 0:
-                    try:
-                        # If we have an address, return it
-                        return codec.to_string(proto2, part)
-                    except Exception as exc:
-                        six.raise_from(exceptions.BinaryParseError(str(exc), self.to_bytes(), proto2.name, exc), exc)
-                else:
-                    # We were given something like '/utp', which doesn't have
-                    # an address, so return ''
-                    return ''
+        """Return the value (if any) following the specified protocol
+        
+        Returns
+        -------
+        union[object, NoneType]
+            The parsed protocol value for the given protocol code or ``None``
+            if the given protocol does not require any value
+        
+        Raises
+        ------
+        ~multiaddr.exceptions.BinaryParseError
+            The stored MultiAddr binary representation is invalid
+        ~multiaddr.exceptions.ProtocolLookupError
+            MultiAddr does not contain any instance of this protocol
+        """
+        proto = protocols.protocol_with_any(proto)
+        for proto2, value in self.items():
+            if proto2 is proto or proto2 == proto:
```

Comment:
Is `proto2 is proto` an optimization because most of the time we use the one imported from `.protocols`, so most of the time(unless we copy it) `proto2 is proto` should work?

Reply:
Yes, it's an optimization as you correctly identified, but it will only be effective for the actual item searched for. All items before that will still call the more complicated  `__eq__`.

For the most part it's just this way here, as I happened to take inspiration from https://github.com/python/cpython/blob/v3.7.3/Lib/_collections_abc.py#L754 when writing this and that has it too. Since the `__eq__` will actually short-circuit when there is a mismatch, I don't think this really amounts to anything.

I can remove this if you prefer a slightly cleaner code.

Reply:
No worries. It's quite interesting and looks good to me! I timeit'ed the one with `is` comparation and the one without it here https://gist.github.com/mhchia/538446d084809dcc0df3dae72d8b1ab6. Though the use case is different, but it gains like 10% with `is` comparation.